Add support for `append` mode for Azure Blob Storage

geovalexis opened the pull request:
### Motivation

Append mode is not a widely used feature in Azure Blob Storage, but it can be particularly useful for specific scenarios, such as writing logs or updating sentinel or marker files. As the native Azure Blob Storage SDK provides support for append mode, this PR integrates that functionality into the `smart_open` interface, making it more accessible and convenient for developers.

This implementation comes with a few limitations that are worth noting:

* There is a maximum size of the blob that can be written in append mode: 195 GiB. This is a limitation of the [Azure Blob Storage API itself](https://learn.microsoft.com/en-us/rest/api/storageservices/understanding-block-blobs--append-blobs--and-page-blobs#about-append-blobs) and cannot be changed.
* The blob must be created in append mode on the first place, meaning that we cannot simply append to an existing blob that may have been created in a different mode (e.g. block or page mode).

As per these limitations, we observe that the append mode is not suitable for all use cases and is not meant to be used indiscriminately by anyone looking to append data to a blob. However, bearing them in mind, many developers may find it convenient for their use cases, especially when they are already using `smart_open` for other operations.

Closes #836

### Tests

New tests have been added to cover the new functionality, including:
- Writing to a new blob in append mode
- Appending to an existing blob in append mode
- Attempting to append to a blob that was not created in append mode (which should raise an error)

Also, I took the chance to fix a few existing tests that were failing when using Azurite. The method `BlobClient._stage_contents` does not exist in the Azure Blob Storage SDK (or at least not in the latest versions), so I replaced it with `BlobClient.get_block_list("uncommitted")`, which returns the same information. 

### Checklist

- [X] Picked a concise, informative and complete title
- [X] Clearly explained the motivation behind the PR
- [X] Linked to any existing issues that your PR will be solving
- [X] Included tests for any new functionality
- [X] Checked that all unit tests pass

geovalexis committed:
fix unit test when running with azurite

geovalexis committed:
add new class for supporting append mode in Azure

geovalexis committed:
add naive approach for azure append

geovalexis committed:
add main append part logic

geovalexis committed:
Merge branch 'develop' of https://github.com/geovalexis/smart_open into feature/azure-append-mode

geovalexis committed:
Merge branch 'develop' into feature/azure-append-mode

geovalexis committed:
fix azure write unit tests when using real blob client

geovalexis committed:
complete Azure append writer

geovalexis committed:
add unit tests for Azure append writer

geovalexis committed:
add more unit tests for Azure's append writer

geovalexis committed:
Merge branch 'develop' of https://github.com/geovalexis/smart_open into feature/azure-append-mode

geovalexis committed:
Merge branch 'develop' of https://github.com/piskvorky/smart_open into feature/azure-append-mode

geovalexis committed:
Merge branch 'develop' of https://github.com/piskvorky/smart_open into feature/azure-append-mode

geovalexis committed:
address linting issues

ddelange commented on smart_open/azure.py:
```diff
@@ -550,3 +558,94 @@ def __repr__(self):
+        """Unsupported."""
+        raise io.UnsupportedOperation
+
+    def detach(self):
+        raise io.UnsupportedOperation("detach() not supported")
```
> why is this raise different from the others?

aha, I see it's copy pasted from above. let's leave it then.

ddelange commented on smart_open/azure.py:
```diff
@@ -550,3 +558,94 @@ def __repr__(self):
+                "input must be one of %r, got: %r" % (_BINARY_TYPES, type(b))
+            )
+
+        # Uploads data as an AppendBlob type with automatic block chunking.
+        # The AppendBlob will be created at first if it does not exist or append to it if it does already.
+        return self._blob.upload_blob(
+            data=b,
+            blob_type=azure.storage.blob.BlobType.APPENDBLOB,
+            overwrite=False,
+            **self._blob_kwargs,
+        )
```
can you implement the `_min_part_size` buffer mechanic similar to the azure `Writer` class? this current implementation is unbuffered, and you'll quickly hit the max parts limit an AppendBlob support if you do many small writes

geovalexis replied:
addressed 👍

geovalexis committed:
enable support for buffered upload to AppendBlob

geovalexis committed:
update azure tests as per new AppendBlob functions

geovalexis committed:
Merge branch 'develop' of https://github.com/piskvorky/smart_open into feature/azure-append-mode

geovalexis committed:
remove unnecessary newlines

geovalexis committed:
fix compression append test as suggested

geovalexis commented on tests/test_azure.py:
```diff
@@ -856,3 +862,138 @@ def test_flush_close(self):
+            with gzip.GzipFile(fileobj=fin) as zipfile:
+                actual = zipfile.read()
+
+        self.assertEqual(expected, actual)
```
addressed 👍

geovalexis commented:
hey! Sorry for the delay on this! it completely fell off my radar, I'm trying to resume my work on this and hopefully get it merged at some point.

geovalexis committed:
fix flake8 linting

geovalexis committed:
Merge branch 'develop' into feature/azure-append-mode

ddelange commented:
@geovalexis is this ready for review?

geovalexis commented:
> @geovalexis is this ready for review?

Yes, it is

ddelange commented on tests/test_azure.py:
```diff
@@ -856,3 +887,141 @@ def test_flush_close(self):
+
+        with self.assertRaises(
+            azure.core.exceptions.ResourceExistsError, msg="The blob type is invalid for this operation."
+        ):
```
noting here that this error is thrown upon write() not upon open(). i think that's analogous to for instance s3 read() on a non-existent key

ddelange commented on tests/test_azure.py:
```diff
@@ -856,3 +887,141 @@ def test_flush_close(self):
+    def test_append_block_over_max_block_size(self):
+        """
+        Does appending into an Azure Blob file work correctly when the block size is over the max block size?
+        By default, this block size is 4MB. Refer to official Azure documentation for more information:
```
conversely, could you add a characterization test for the min_part_size mechanic? ie first write of min_part_size - 1 doesnt trigger upload, then writing 1 byte triggers upload? 

_DEFAULT_MIN_PART_SIZE is set to 64MB in azure.py so that edge case is already implicitly covered. make it explicit? some way we can hit this 4MB limit or is it completely abstracted away for the smart_open end-user?

geovalexis replied:
Updated code with new requested unit test 👍 . About 4Mb limit, that's completely abstracted away from smart-open, Azure SDK's handled this automatically by splitting the data into 4Mb chunks if necessary.

ddelange reviewed (commented):
i like the test coverage! can you add a note to the end of [this](https://github.com/piskvorky/smart_open?tab=readme-ov-file#azure-advanced-usage) README section mentioning mode="ab"? no example needed, just mention the functionality?

geovalexis committed:
add unit test for append min part size on azure and update azure docs accordingly

geovalexis committed:
update broken link on azure unit tests

geovalexis commented:
@ddelange thanks for reviewing my PR and your comments. I've pushed new changes addressing your comments, please do not hesitate to push back if there's something I missed.

ddelange commented on README.rst:
```diff
@@ -467,6 +467,8 @@ Additional keyword arguments can be propagated to the ``commit_block_list`` meth
     kwargs = {'metadata': {'version': 2}}
     fout = open('azure://container/key', 'wb', transport_params={'blob_kwargs': kwargs})
 
+``smart_open`` also supports append mode (``mode="ab"``) for Azure Blob Storage using `Append Blobs <https://learn.microsoft.com/en-us/rest/api/storageservices/understanding-block-blobs--append-blobs--and-page-blobs#about-append-blobs>`__. Note that the blob must have been created as an Append Blob and is subject to a maximum size of 195 GiB.
```
```suggestion
For Azure, ``smart_open`` also supports append mode (``mode="ab"``) by creating new (or appending to existing) `Append Blobs <https://learn.microsoft.com/en-us/rest/api/storageservices/understanding-block-blobs--append-blobs--and-page-blobs#about-append-blobs>`__ (max 195 GiB).
```

ddelange committed:
Make README language more concise

ddelange reviewed (approved):
@geovalexis many thanks for your contribution!

ddelange committed:
Add lower version pin and update docstrings
